l2geth/rollup/sync_service.go

```diff
@@ -763,10 +680,211 @@ func (s *SyncService) ApplyTransaction(tx *types.Transaction) error {
+			return nil
+		}
+		if err := s.syncTransactionRange(nextIndex, *latestIndex, backend); err != nil {
+			return err
+		}
+
+		post, err := s.client.GetLatestTransaction(backend)
+		if err != nil {
+			return fmt.Errorf("Cannot get latest transaction: %w", err)
+		}
+		postLatestIndex := post.GetMeta().Index
+		if postLatestIndex == nil {
+			return errors.New("Latest index is nil")
+		}
+
+		if *postLatestIndex == *latestIndex {
+			return nil
+		}
+	}
+}
+
+// syncTransactionRange will sync a range of transactions from
+// start to end (inclusive) from a specific Backend
+func (s *SyncService) syncTransactionRange(start, end uint64, backend Backend) error {
+	log.Info("Syncing transactions to tip", "start", start, "end", end)
+	for i := start; i <= end; i++ {
+		tx, err := s.client.GetTransaction(i, backend)
+		if err != nil {
+			return fmt.Errorf("cannot fetch transaction %d: %w", i, err)
+		}
+		if err = s.applyTransaction(tx); err != nil {
+			return fmt.Errorf("Cannot apply transaction: %w", err)
+		}
+	}
+	return nil
+}
+
+// syncTransactionBatchesToTip will sync all of the transaction batches to the
+// tip
+func (s *SyncService) syncTransactionBatchesToTip() error {
+	s.loopLock.Lock()
+	defer s.loopLock.Unlock()
+	log.Debug("Syncing transaction batches to tip")
+
+	for {
+		latest, _, err := s.client.GetLatestTransactionBatch()
+		if errors.Is(err, errElementNotFound) {
+			log.Info("No transaction batches to sync")
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("Cannot get latest transaction batch: %w", err)
+		}
+
+		nextIndex := s.GetNextBatchIndex()
+		if nextIndex-1 == latest.Index {
+			log.Info("No new batches to sync", "tip", latest.Index)
+			return nil
+		}
+		if err := s.syncTransactionBatchRange(nextIndex, latest.Index); err != nil {
+			return err
+		}
+
+		post, _, err := s.client.GetLatestTransactionBatch()
+		if err != nil {
+			return fmt.Errorf("Cannot get latest transaction batch: %w", err)
+		}
+		if post.Index == latest.Index {
+			return nil
+		}
+	}
+}
+
+// syncTransactionBatchRange will sync a range of batched transactions from
+// start to end (inclusive)
+func (s *SyncService) syncTransactionBatchRange(start, end uint64) error {
+	log.Debug("Syncing transaction batch range", "start", start, "end", end)
+	for i := start; i <= end; i++ {
+		log.Debug("Fetching transaction batch", "index", i)
+		_, txs, err := s.client.GetTransactionBatch(i)
+		if err != nil {
+			return fmt.Errorf("Cannot get transaction batch: %w", err)
+		}
+		for _, tx := range txs {
+			if err := s.applyBatchedTransaction(tx); err != nil {
+				return fmt.Errorf("cannot apply batched transaction: %w", err)
+			}
+		}
+		s.SetLatestBatchIndex(&i)
+	}
+	return nil
+}
+
+// syncQueueToTip will sync the `enqueue` transactions to the tip
+// from the last known `enqueue` transaction
+func (s *SyncService) syncQueueToTip() error {
+	s.loopLock.Lock()
+	defer s.loopLock.Unlock()
+
+	for {
```

Comment:
Please refactor these infinite loops into single iteration functions which can be individually unit tested. The function should then be called inside the loop.